File: 17 - Blood Alcohol Calculator/Golang/main.go
package main

import (
    "fmt"
    "strconv"
)

func input_float(output string) float64 {
    hold := "NaN"
    for {
        fmt.Printf(output)
        fmt.Scanf("%s", &hold)

        f, err := strconv.ParseFloat(hold, 64)
        if err != nil {
            continue
        } else {
            return f
        }
    }
}

func calculate_BAC(A float64, W float64, r float64, H float64) float64 {
    return (A * 5.14/W  * r) - 0.015 * H;
}

func main() {
    alcohol := input_float("How much alcohol have you consumed? ")
    weight  := input_float("What is your weight in pounds? ")
    hours   := input_float("How many hours ago was your last drink? ")

    var bac float64 = 0

    var sex string  = "NaS"
    MALE           := []string{"male", "man", "m"}
    FEMALE         := []string{"female", "woman", "f"}
    for ( !in_string_slice(sex, MALE) && !in_string_slice(sex, FEMALE) ) {
        fmt.Printf("Is your biological sex male or female? ")
        fmt.Scanf("%s", &sex)
    }

    if ( in_string_slice(sex, MALE) ) {
        bac = calculate_BAC(alcohol, weight, 0.73, hours)
        fmt.Printf("Your blood alcohol content is %f\n", bac)
    } else {
        bac = calculate_BAC(alcohol, weight, 0.66, hours)
        fmt.Printf("Your blood alcohol content is %f\n", bac)
    }

    if ( bac >= 0.08 ) { fmt.Printf("It is not legal for you to drive.\n") }
    if ( bac <= 0.08 ) { fmt.Printf("It is legal for you to drive.\n")     }

}

func in_string_slice(element string, list []string) bool {
    for _, val := range list {
        if ( val == element ) {
            return true
        } else {
            continue
        }
    }
    return false // element was not found in list
}

File: 17 - Blood Alcohol Calculator/C++/main.cpp
#include <iostream>
#include <algorithm>

using namespace std;

float get_float(string output)
{
    float temp = 0;
    cout << output;
    cin >> temp;
    return temp;
}

bool string_in_array(string s , string a[], int size) {
    for (int i = 0; i < size; i++) {
        if (a[i] == s) {
            return true;
        }
    }
    return false;
}

float calculate_bac(float A, float W, float r, float H)
{
    return ( A * 5.14/W * r ) - 0.015 * H;
}

int main(void)
{
    float alcohol, weight, hours, bac = 0;
    string MALE[3] = { "m", "male", "man" };
    string FEMALE[3] = { "f", "female", "woman" };

    alcohol = get_float("How much alcohol have you consumed in oz.? ");
    weight  = get_float("How much do you weight? ");
    hours   = get_float("How many hours ago was your last drink? ");

    while ( 1 ) {
        string sex = "";
        cout << "What is your biological sex? ";
        cin >> sex;
        transform(sex.begin(), sex.end(), sex.begin(), ::tolower);
        if ( string_in_array(sex, MALE, sizeof(MALE)) ) {
            bac = calculate_bac(alcohol, weight, 0.73, hours);
            break;
        } else if ( string_in_array(sex, FEMALE, sizeof(FEMALE)) ) {
            bac = calculate_bac(alcohol, weight, 0.66, hours);
            break;
        }
    }

    cout << "Your BAC is: " << bac << endl;

    if (bac >= 0.08) {
        cout << "It is not legal for you to drive." << endl;
    } else {
        cout << "It is legal for you to drive." << endl;
    }

    return 0;
}
